README.md
Burst Detector for Preterm EEG
==============================

Collection of M-files (computer code) to implement a method to detect bursts on EEG
recordings, as described in [[1]](#references). Detector uses a linear combination of
multiple amplitude and spectral features.  Developed and tested on EEG recordings from
premature infants (<30 weeks gestational age) using a bipolar EEG montgage (F3-C3, F4-C4,
C3-O1, C4-O2, C3-T3, C4-T4, Cz-C3, and C4-Cz). Detector operates on 1 channel only.
Requires Matlab or Octave programming environments. Updates can be found
at [github](https://github.com/otoolej/burst_detector).

To cite this software, please use reference [[1]](#references). 

__Update (Nov. 2019): Python version of this code at:__ https://github.com/otoolej/py_burst_detector

[![DOI](https://zenodo.org/badge/42042482.svg)](https://zenodo.org/badge/latestdoi/42042482)

# contents
* [overview](#overview)
* [quick start](#quick-start)
* [example](#example)
* [requirements](#requirements)
* [test computer setup](#test-computer-setup)
* [licence](#licence)
* [references](#references)
* [contact](#contact)


# overview

![Example of burst detector (purple annotation) on channel F3-C3](pics/detector_example_for_website.png)

A simple method to detect bursts in the EEG of preterm infants. The method was developed
by assessing multiple frequency and amplitude features of bursts. Selected features were combined in
a classifier (support vector machine). After a feature selection and training procedure,
the detector consisted of eight features which are combined in a linear support vector
machine. The code here implements this detector, which was trained on annotations from
1-channel of 10 minute EEG recordings from 36 preterm infants.

![Another example of burst detector on channel C4-T4](pics/detector_example_for_website_v2.png)


# quick start
Set paths in Matlab/Octave, or do so using the `load_curdir` function:
```matlab
  >> load_curdir;
```


# example
```matlab
      % use impulsive noise test signal
      N=5000;  Fs=64; 
      x=gen_impulsive_noise(N).*10;
      
	  % detect the bursts:
      [burst_anno,svm_out]=eeg_burst_detector(x,Fs);

      % and plot:
      figure(1); clf; 
      t=(0:N-1)./Fs;
      hax(1)=subplot(211); hold all;
      plot(t,burst_anno); plot(t,svm_out);
      hax(2)=subplot(212); 
      plot(t,x);
      linkaxes(hax,'x');
      xlabel('time (seconds)');
```

![Burst detector (in Matlab environment) for impulsive noise test signal](pics/burst_detector_impulsive_noise_example.png)

# files
All Matlab files (.m files) have a description and an example in the header. To read this
header, type `help <filename.m>` in Matlab.  Directory structure is as follows: 
```
├── bd_parameters.m               # parameters for the method are defined here
├── eeg_burst_detector.m          # burst detector for multi-channel EEG
├── estimate_features
│   ├── edo.m                       # envelope--derivative operator
│   ├── env.m                       # signal envelope (using Hilber transform)
│   ├── fd.m                        # fractal dimension (using Higuichi method)
│   ├── if_.m                       # instantaneous frequency (using CFD estimate)
│   ├── psd_r2.m                    # fit of linear regression to log-log spectrum
│   └── rel_spec_power.m            # relative spectral power
├── gen_features.m                  # generate features
├── load_curdir.m                   # load the path
└── utils
    ├── do_bandpass_filtering.m     # bandpass filtering (IIR filter)
    ├── epoch_window.m              # epoch window size and hop lengths
    ├── get_window.m                # generate epoch window
    └── gen_impulsive_noise.m       # impulsive noise
```


# requirements
Either Matlab (R2013 or
newer, [Mathworks website](http://www.mathworks.co.uk/products/matlab/)) with the Signal
Processing Toolbox or Octave (v3.8 or
newer, [Octave website](http://www.gnu.org/software/octave/index.html)) with the
'octave-signal' add-on package.



# test computer setup
- hardware:  Intel(R) Xeon(R) CPU E5-1603 0 @ 2.80GHz; 8GB memory.
- operating system: Ubuntu GNU/Linux x86_64 distribution (16.10, Yakkety Yak), with Linux
  kernel 4.8.0-41-generic.
- software: Octave (4.0.3) with 'octave-signal' toolbox and Matlab (R2013a) with Signal
  Processing Toolbox.

---

# licence

```
Copyright (c) 2015, John M. O' Toole, University College Cork
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

  Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

  Redistributions in binary form must reproduce the above copyright notice, this
  list of conditions and the following disclaimer in the documentation and/or
  other materials provided with the distribution.

  Neither the name of the University College Cork nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```


# references

1. JM O' Toole, GB Boylan, RO Lloyd, RM Goulding, S Vanhatalo, and NJ Stevenson,
“Detecting Bursts in the EEG of Very and Extremely Premature Infants Using a Multi-Feature
Approach”, Medical Engineering & Physics, vol. 45, pp. 42-50, 2017. 
[DOI:10.1016/j.medengphy.2017.04.003](https://doi.org/10.1016/j.medengphy.2017.04.003)
(Version of code to accompany paper: [![DOI](https://zenodo.org/badge/42042482.svg)](https://zenodo.org/badge/latestdoi/42042482))


2. JM O' Toole and NJ Stevenson, “Assessing instantaneous energy in the EEG: a
non-negative, frequency-weighted energy operator”, In 36th Annual International Conference
of the IEEE Engineering in Medicine and Biology Society (EMBC), pp. 3288-3291, 2014. 
{ [paper](http://otoolej.github.io/pubs/PDFs/JMOToole_energy_EMBC14.pdf) |
  [poster](http://otoolej.github.io/pubs/PDFs/EMBC_poster_Aug2014_NLEO.pdf) |
  [code](http://otoolej.github.io/code/nleo/) }


---

# contact

John M. O' Toole

Neonatal Brain Research Group,  
Irish Centre for Fetal and Neonatal Translational Research ([INFANT](http://www.infantcentre.ie/)),  
Department of Paediatrics and Child Health,  
University College Cork,  
Cork University Hospital, Room 2.19 Paediatrics Building,  
Cork, Ireland




CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. Project started in
2015-26-06 as single-channel detector.  This log started 2017-03-20, after version
0.1.0. More details can be found in the git logs.

## unreleased
### Fix
- bug in edo.m for Fs_new / Fs not an integer

## 0.1.1
### Changed 
- not using 'polyfit' to reduce computational time
- include NaN's in output annotation

### Added
- this CHANGELOG.md!
- now using semantic versioning (see http://semver.org/)
- PDF version of README in addition to markdown version



gen_features.m
%-------------------------------------------------------------------------------
% gen_features: generate feature set for signal x
%
% Syntax: [t_stat]=gen_features(x,Fs,feat_set)
%
% Inputs: 
%     x        - 1-channel of EEG data (size 1 x N)
%     Fs       - sampling frequency (in Hz, must be ≥64 Hz)
%     feat_set - feature set ('FEATURE_SET_FINAL' in bd_parameters.m, size 1 x P)
%
% Outputs: 
%     t_stat   - features estimated from input signal x (size P x N)
%
% Example:
%     N=5000; Fs=64; 
%     x=gen_impulsive_noise(N).*10;
%
%     t_stat=gen_features(x,Fs);
% 
%     figure(1); clf; 
%     t=(0:N-1)./Fs;
%     hax(1)=subplot(211);
%     plot(t,t_stat'); ylim([-5,5]);
%     hax(2)=subplot(212); 
%     plot(t,x);
%     linkaxes(hax,'x');
%     xlabel('time (seconds)');


% John M. O' Toole, University College Cork
% Started: 17-06-2015
%
% last update: Time-stamp: <2015-06-26 10:06:21 (otoolej)>
%-------------------------------------------------------------------------------
function [t_stat]=gen_features(x,Fs,feat_set)
if(nargin<2), error('input arguments should include x and Fs.'); end
if(nargin<3 || isempty(feat_set)), feat_set=[]; end


N=length(x);


%---------------------------------------------------------------------
% 0. set parameters
%---------------------------------------------------------------------
bd_parameters;

% do log-transform for these features :
FEAT_LOG_TRANSFORM={'edo','envelope','spec-power' };


% feature set
if(isempty(feat_set))
    feat_set=FEATURE_SET_FINAL;
end
N_feats=length(feat_set);

t_stat=NaN(N_feats,N);

% for missing data, insert 0's when generating the features
% and then, afterwards, replace with NaNs.
inans=find(isnan(x));
x(inans)=0;


%---------------------------------------------------------------------
% 1. do band-pass filtering first
%---------------------------------------------------------------------
filter_bands={[0 1 0 0 0], [0 0 1 0 0], [0 0 0 1 0], [0 0 0 0 1]};
x_filt=[];
for n=1:length(filter_bands)
    if( any(cellfun(@(x) isequal(filter_bands{n},x), {feat_set.params})) )
        x_filt{n}=filt_b(x,Fs,FBANDS(n+1,2),FBANDS(n+1,1),L_FILTER_ORDER);
    else
        x_filt{n}=[];
    end
end




% 1. iterate over all features:
for n=1:N_feats
    
    y=[];
    if(~isempty(feat_set(n).params))
        ifilt=find(cellfun(@(x) isequal(FEATURE_SET_FINAL(n).params,x), filter_bands));
        y=x_filt{ifilt};
    end
    
    %---------------------------------------------------------------------
    % 2. calculate features
    %---------------------------------------------------------------------
    switch feat_set(n).name
      case {'spec-power'}
        %---------------------------------------------------------------------
        % relative spectral power
        %---------------------------------------------------------------------
        p=find(feat_set(n).params);
        t_stat(n,:)=rel_spec_power(x,Fs,FBANDS(p,:),FBANDS(1,:));
        
      case 'psd_r2'
        %---------------------------------------------------------------------
        % fit of line to log-log PSD
        %---------------------------------------------------------------------
        p=find(feat_set(n).params);        
        t_stat(n,:)=psd_r2(y,Fs,FBANDS(p,:));
        
      case 'if'
        %---------------------------------------------------------------------
        % instantaneous frequency
        %---------------------------------------------------------------------
        p=find(feat_set(n).params);
        t_stat(n,:)=if_(y,Fs,FBANDS(p,:));
        
      case 'envelope'
        %---------------------------------------------------------------------
        % envelope (Hilbert transform)
        %---------------------------------------------------------------------
        t_stat(n,:)=env(y,Fs);

      case 'edo'
        %---------------------------------------------------------------------
        % envelope--derivative operator
        %---------------------------------------------------------------------
        t_stat(n,:)=edo(x,Fs);
        
      case 'fd-higuchi'
        %---------------------------------------------------------------------
        % fractal dimension estimate (Higuchi method)
        %---------------------------------------------------------------------
        t_stat(n,:)=fd(x,Fs);
        
        
      otherwise
        error(['unknown feature: ' feat_set(n).name]);
    end

    
    if(any(ismember(FEAT_LOG_TRANSFORM,feat_set(n).name)))
        t_stat(n,:)=log( t_stat(n,:) + eps );
    end
    
    
end

    
% for missing data:
t_stat(:,inans)=NaN;






function y=filt_b(x,Fs,F3db_lowpass,F3db_highpass,order)
%---------------------------------------------------------------------
% IIR zero-phase filter (Butterworth)
%---------------------------------------------------------------------
if(nargin<2 || isempty(Fs)), Fs=1; end
if(nargin<3 || isempty(F3db_lowpass) || F3db_lowpass==0), F3db_lowpass=[]; end
if(nargin<4 || isempty(F3db_highpass) || F3db_highpass==0), F3db_highpass=[]; end
if(nargin<5 || isempty(order)), order=3; end


if(isempty(F3db_highpass))
    [b,a]=butter(order,F3db_lowpass/(Fs/2),'low');    
    
elseif(isempty(F3db_lowpass))
    [b,a]=butter(order,F3db_highpass/(Fs/2),'high');        
    
else
    y=filt_b(x,Fs,F3db_lowpass,[],order);    
    y=filt_b(y,Fs,[],F3db_highpass,order);        
    return;    
end


y=filtfilt(b,a,x);











